### DISCUSSION TITLE
Memory Optimisation & const()

### DISCUSSION THREAD

**Discussion by Beormund** [micropython/micropython#10119] [Using MicroPython]
If I have a lookup like so:
```py
lookup = {
    'on': 1,
    'off': 0,
    '1': 1,
    '0': 0,
    'true': 1,
    'false': 0
}
```
How can I best utilise const() to optimise RAM use? Do I first have to declare _on = const('on') and then use _on as a dictionary key? Same with values?

With regards to long strings, would there be any advantage to declaring XML strings like so:
```py
_GETBINARYSTATE_SOAP = const("""<s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/" s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/">
  <s:Body>
    <u:GetBinaryStateResponse xmlns:u="urn:Belkin:service:basicevent:1">
      <BinaryState>%(state_relay)s</BinaryState>
    </u:GetBinaryStateResponse>
  </s:Body>
</s:Envelope>
""")
```
And is this the correct way to declare a const tuple?
```py
_abbr = const(('last', 'first', 'second', 'third', 'fourth'))
```

**Comment by peterhinch** [comment]
The `const` declaration only works when used in global scope e.g.
```python
_t = const((1, 2, 3))
```
Without the `const` declaration the compiler would generate a global variable `_t` with the value `(1, 2, 3)`. For any reference to `_t` code would be generated to look up the variable `_t`. With the `const` declaration, a reference such as
```python
foo(_t)
```
would be compiled as
```python
foo((1, 2, 3))
```
In other words, `_t` would use no RAM because the symbol is expanded at compile time; the time used to access the global would also be saved. However the constant tuple `(1, 2, 3)` has to be stored somewhere: as a `const` it is stored once in code space. This will be either RAM or flash depending on whether the code is frozen as bytecode.
